### PR TITLE
[webview_flutter] Add getters and setters for cookies.

### PR DESCRIPTION
**Pull request by jeroen-meijer**
## Description

This PR adds better support for cookies by adding `getCookies` and `setCookies` methods to the `CookieManager`.

Instead of a `List<String>`, these methods will use a `List<Cookie>` where the `Cookie` class comes from `dart:io`. This provides the user with easier cookie handling because the cookies from the `WebView` instance no longer have to be manually deserialised from `String`s. Also, this provides better interop with Dart's native `HttpClient`.

## Related Issues

This PR may help solve the following issues:
flutter/flutter#27795
flutter/flutter#27624
flutter/flutter#27597

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by amirh** [comment]
Thanks for the contribution!

I'm following the [initial PR review policy](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#the-review-process), this PR isn't trivial to review so I'm labeling it with "backlog" and we will prioritize according to the priority of the associated issues.

**Comment by joelbrostrom** [comment]
Status on this PR?
Need it :)

**Comment by amirh** [comment]
@joelbrostrom see the comment above - the issues will be prioritized as described in the [wiki](https://github.com/flutter/flutter/wiki/Issue-hygiene#prioritization).

**Comment by 18316603765** [comment]
How to use it

**Comment by 18316603765** [comment]
现在更新版本里面没setcookie 呀，什么时候发布新版本

**Comment by jeroen-meijer** [comment]
@joelbrostrom @18316603765 
Don't forget you can already use this PR by adding `webview_flutter` to your projects by adding this to your `pubspec.yaml`:
```yaml
  webview_flutter:
    git: 
      url: git://github.com/devangels/plugins.git
      ref: cookies
      path: packages/webview_flutter
```

This does mean that any updates for `webview_flutter` done after this PR was submitted aren't there, so use it with caution.

**Comment by Pekno** [comment]
Any news on it being added soon ? I can't seem to use the git path.

**Comment by Pekno** [comment]
I managed to add the package to my app, I needed to replace the 'git://github.com/devangels/plugins.git' to 'https://github.com/devangels/plugins.git' for it to work.
But now I can't seem to build my app anymore.

`What went wrong:Execution failed for task ':webview_flutter:compileDebugJavaWithJavac'.
 Unable to find source java class: 'C:\Users\1604742\Documents\flutter\.pub-cache\hosted\pub.dartlang.org\webview_flutter-0.3.13\android\src\main\java\io\flutter\plugins\webviewflutter\DisplayListenerProxy.java' because it does not belong to any of the source dirs: '[C:\Users\1604742\AppData\Roaming\Pub\Cache\git\plugins-c73e8b7fd115fcdeaf512239905337017f525971\packages\webview_flutter\android\src\main\java, C:\Users\1604742\AppData\Roaming\Pub\Cache\git\plugins-c73e8b7fd115fcdeaf512239905337017f525971\packages\webview_flutter\android\src\debug\java, C:\Users\1604742\Documents\Projet\flutter_app\build\webview_flutter\generated\not_namespaced_r_class_sources\debug\generateDebugRFile\out, C:\Users\1604742\Documents\Projet\flutter_app\build\webview_flutter\generated\source\buildConfig\debug, C:\Users\1604742\Documents\Projet\flutter_app\build\webview_flutter\generated\source\aidl\debug, C:\Users\1604742\Documents\Projet\flutter_app\build\webview_flutter\generated\source\rs\debug]'`

**Comment by ashbin** [comment]
@jeroen-meijer can you merge this please.

**Comment by twopai** [comment]
most of flutter webview, their all not method for the setcookies ,but communicate with js by it ,so can you add the api directly！
and today , i do the setCookies method on my program by your CookieManager, but not work ,help me thanks !

**Comment by twopai** [comment]
most of flutter webview, their all not method for the setcookies ,but communicate with js by it ,so can you add the api directly！
and today , i do the setCookies method on my program by your CookieManager, but not work ,help me thanks !

**Comment by twopai** [comment]
most of flutter webview, their all not method for the setcookies ,but communicate with js by it ,so can you add the api directly！
and today , i do the setCookies method on my program by your CookieManager, but not work ,help me thanks !

**Comment by jeroen-meijer** [comment]
@ashbin I can't, I'm don't have the permissions.
@twopai Wait for this PR to land before reporting error in specific workflows.

**Comment by kwongfung** [comment]
Need it

File: packages/webview_flutter/ios/Classes/FLTCookieManager.m [hunk 1]
```diff
@@ -1,49 +1,69 @@
...
+    result(serialized);
+  }];
+}
+
+- (void)setCookies:(FlutterMethodCall *)call result:(FlutterResult)result API_AVAILABLE(ios(11.0)) {
+  NSArray<CookieDto *> *cookieDtos = [CookieDto manyFromDictionaries:[call arguments]];
+  for (CookieDto *cookieDto in cookieDtos) {
+    [cookieStore setCookie:[cookieDto toNSHTTPCookie]
+         completionHandler:^(){
+         }];
```
Comment (wwwdata) [inline packages/webview_flutter/ios/Classes/FLTCookieManager.m]: I tested this out and found several issues:

- the `result` handler will never be called => your async dart code is stuck, same for `clearCookies`
- the cookie data only considers `name` and `value`. So this is not going to work at all. In order for cookies to work we also need to consider other properties like the `domain`. But after adding the rest of the fields, and implementing some stupid counter code, I made it work.

Stupid counter code that calls `result` in the end example:

```objc
- (void)setCookies:(FlutterMethodCall *)call result:(FlutterResult)result API_AVAILABLE(ios(11.0)) {
    NSArray<CookieDto *> *cookieDtos = [CookieDto manyFromDictionaries:[call arguments]];
    __block int counter = 0;
    for (CookieDto *cookieDto in cookieDtos) {
        [cookieStore setCookie:[cookieDto toNSHTTPCookie]
             completionHandler:^(){
            counter++;
            if (counter == cookieDtos.count) {
                result(nil);
            }
        }];
    }
}
```

File: packages/webview_flutter/ios/Classes/CookieDto.m [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  return @{
+    @"name" : name,
+    @"value" : value,
+  };
+};
+
+- (NSHTTPCookie *)toNSHTTPCookie {
+  return [NSHTTPCookie cookieWithProperties:@{NSHTTPCookieName : name, NSHTTPCookieValue : value}];
+}
+
```
Comment (wwwdata) [inline packages/webview_flutter/ios/Classes/CookieDto.m]: Other properties like 

- `NSHTTPCookieDomain`
- `NSHTTPCookieSecure`
- `NSHTTPCookiePath`
 
are missing. After adding them, it worked for me

**Comment by jeroen-meijer** [comment]
Thanks, @wwwdata! I did the best I could at the time, but I really appreciate you stepping in and taking a look.

I'll take a look at the changes you requested.

File: packages/webview_flutter/ios/Classes/CookieDto.m (hunk 2 above)
Comment (jeroen-meijer) [reply packages/webview_flutter/ios/Classes/CookieDto.m]: Got it. Do you propose those fields should be added to the CookieDto (in Dart, Java and ObjC) as well then?

File: packages/webview_flutter/ios/Classes/FLTCookieManager.m (hunk 1 above)
Comment (jeroen-meijer) [reply packages/webview_flutter/ios/Classes/FLTCookieManager.m]: Thanks. I'll use the code you provided until I or someone else comes up with a better solution.

File: packages/webview_flutter/ios/Classes/CookieDto.m (hunk 2 above)
Comment (wwwdata) [reply packages/webview_flutter/ios/Classes/CookieDto.m]: Yes, I would support all the fields that the `Cookie` class in Dart has.

**Comment by wwwdata** [comment]
Good changes, I think sending it as a string makes the most sense for Android. I now also found out that the `toString` method when you construct the `Cookie` object in Android is buggy and my open PR does not work.

I have to check out your iOS part as well. I noticed that using this process pool was necessary to cookies sync across different web views. If you would also use that and everything works, I think we should push this PR through and close the other one from me again. You can check the code on my PR for how to do the process pool in iOS.

**Comment by jeroen-meijer** [comment]
@googlebot I consent.

**Comment by aboutmydreams** [comment]
Error output from CocoaPods:
↳
        WARNING: CocoaPods requires your terminal to be using UTF-8 encoding.
        Consider adding the following to ~/.profile:
        export LANG=en_US.UTF-8
        
    [!] Automatically assigning platform `iOS` with version `8.0` on target `Runner` because no platform was specified. Please specify a platform for this target in your Podfile. See `https://guides.cocoapods.org/syntax/podfile.html#platform`.
Error running pod install
Error launching application on iPhone 11 Pro Max.
Exited (sigterm)

**Comment by dotcink** [comment]
Is this still in progress?
If not, I might submit another PR.

**Comment by wwwdata** [comment]
> Is this still in progress?
> If not, I might submit another PR.

what would be different in your PR? Did you also have a look at my open PR? It is a bid sad that this PR is now open for so long since cookie syncing is, in my opinion, a very important topic.

**Comment by dotcink** [comment]
> > Is this still in progress?
> > If not, I might submit another PR.
> 
> what would be different in your PR? Did you also have a look at my open PR? It is a bid sad that this PR is now open for so long since cookie syncing is, in my opinion, a very important topic.

Nothing special.
I wrote `setCookie` feature before searched this PR, but its checks seem having failed for more than a month, so I asked the progress.

**Comment by teejayhh** [comment]
This is something we could use, and infact its a requirement. Please

**Comment by Justwen** [comment]
Could this PR be available during my rest of my life ?

**Comment by wwwdata** [comment]
@Justwen maybe, maybe not, who knows! Going to take some time this week to rebase this again.. and then let's wait.. again, and hope!

**Comment by cklee75** [comment]
This is important feature for webview_flutter plugin as the existing solution which use Javascript eval document.cookie cannot retrieve secured/HttpOnly cookies, which is important for certain authentication solution.

**Comment by joaoarmando** [comment]
Any news about it?

**Comment by fryette** [comment]
If someone needed a solution
https://pub.dev/packages/webview_cookie_manager

**Comment by rodruiz** [comment]
> If someone needed a solution
> https://pub.dev/packages/webview_cookie_manager

Confirming this package works.

**Comment by jeroen-meijer** [comment]
Though I also believe that cookie handling is important in Flutter, I currently don't have more time to spend on this PR. 😔

The changes are still available on the source branch, though, so if anybody wants to fork it, pick this up and continue working on it, be my guest.

In the meantime, the package proposed by @fryette seems to work, so I recommend people to use that instead.

Closing this. Keeping the branch open.

**Pull request closed without merging**